### PR TITLE
allow commissionee device read fabric list in Android

### PR DESCRIPTION
**Pull request by nicelyjust**
#### Problem
On some devices with a screen（such as TV）, User want to know which commissioners currently have and how many，but now we dont support it.

#### Change overview
allows the commissionee to see what the current commissioners are，so add two JNI function，one reads fabric count，the other reads fabric list.

#### Testing
- run ./scripts/examples/gn_build_example.sh examples/chip-tool

- run ./scripts/examples/gn_android_example.sh examples/tv-app/android out/android-arm64-chip-tvserver

- chip-tool pairing

### CODE REVIEW

File: src/app/server/java/ChipFabricProvider-JNI.cpp [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+{
+    jobject jFabricList;
+    JniReferences::GetInstance().CreateArrayList(jFabricList);
+    jclass jFabricCls = env->FindClass("chip/appserver/Fabric");
+    VerifyOrExit(jFabricCls != nullptr, ChipLogError(NotSpecified, "unfind Class Fabric"));
+    {
+        for (auto & fabricInfo : Server::GetInstance().GetFabricTable())
+        {
+            // see src/app/clusters/operational-credentials-server/operational-credentials-server.cpp#ReadFabricsList
+            if (!fabricInfo.IsInitialized())
```
Comment (bzbarsky-apple) [inline src/app/server/java/ChipFabricProvider-JNI.cpp:98]: Please see review comments on this in the other PR where this code was proposed.

File: src/app/server/java/ChipFabricProvider-JNI.cpp [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    ChipLogProgress(DeviceLayer, "ChipFabricProvider JNI_OnLoad() called");
+
+    chip::Platform::MemoryInit();
+
+    // Save a reference to the JVM.  Will need this to call back into Java.
+    JniReferences::GetInstance().SetJavaVm(jvm, "chip/appserver/ChipFabricProvider");
+    sJVM = jvm;
+
+    // Get a JNI environment object.
+    env = JniReferences::GetInstance().GetEnvForCurrentThread();
```
[the remark this answers is not in the document]
Comment (nicelyjust) [reply src/app/server/java/ChipFabricProvider-JNI.cpp:60]: thank for the suggestion，just check the JNI environment is right

File: src/app/server/java/ChipFabricProvider-JNI.cpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    VerifyOrExit(env != NULL, err = CHIP_JNI_ERROR_NO_ENV);
+
+    ChipLogProgress(DeviceLayer, "Loading Java class references.");
+
+    chip::InitializeTracing();
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        // JniReferences::GetInstance().ThrowError(env, sFabricProviderCls, err);
```
[the remark this answers is not in the document]
Comment (nicelyjust) [reply src/app/server/java/ChipFabricProvider-JNI.cpp:70]: OK, i'll remove it

File: src/app/server/java/ChipFabricProvider-JNI.cpp [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+    // Save a reference to the JVM.  Will need this to call back into Java.
+    JniReferences::GetInstance().SetJavaVm(jvm, "chip/appserver/ChipFabricProvider");
+    sJVM = jvm;
+
+    // Get a JNI environment object.
+    env = JniReferences::GetInstance().GetEnvForCurrentThread();
+    VerifyOrExit(env != NULL, err = CHIP_JNI_ERROR_NO_ENV);
+
+    ChipLogProgress(DeviceLayer, "Loading Java class references.");
```
[the remark this answers is not in the document]
Comment (nicelyjust) [reply src/app/server/java/ChipFabricProvider-JNI.cpp:63]: OK, I'll remove it.

File: src/app/server/java/ChipFabricProvider-JNI.cpp [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+{
+    return chip::Server::GetInstance().GetFabricTable().FabricCount();
+}
+
+jobject ReadFabricList(JNIEnv * env, jobject self)
+{
+    jobject jFabricList;
+    JniReferences::GetInstance().CreateArrayList(jFabricList);
+    jclass jFabricCls = env->FindClass("chip/appserver/Fabric");
+    VerifyOrExit(jFabricCls != nullptr, ChipLogError(NotSpecified, "unfind Class Fabric"));
```
[the remark this answers is not in the document]
Comment (nicelyjust) [reply src/app/server/java/ChipFabricProvider-JNI.cpp:93]: sorry, wrong verb,I'll correct it

File: src/app/server/java/ChipFabricProvider-JNI.cpp (hunk 1 above)
Comment (nicelyjust) [reply src/app/server/java/ChipFabricProvider-JNI.cpp:98]: @bzbarsky-apple Hi, Boris, not found any comments about it in  https://github.com/project-chip/connectedhomeip/pull/12019, could you help to check this?

Comment (bzbarsky-apple) [reply src/app/server/java/ChipFabricProvider-JNI.cpp]: Why are you looking in #12019?  See https://github.com/project-chip/connectedhomeip/pull/18546#discussion_r875958372

Comment (nicelyjust) [reply src/app/server/java/ChipFabricProvider-JNI.cpp]: Sorry, I misunderstand , @bzbarsky-apple I refer to src/app/clusters/operational-credentials-server/operational-credentials-server.cpp#ReadFabricsList function, it does the same thing, is it not necessary to do this?

File: src/app/server/java/ChipFabricProvider-JNI.cpp [hunk 6]
```diff
@@ -0,0 +1,129 @@
+/*
+ *   Copyright (c) 2021 Project CHIP Authors
```
Comment (andy31415) [inline src/app/server/java/ChipFabricProvider-JNI.cpp:2]: nit: 2022

File: src/app/server/java/ChipFabricProvider-JNI.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (andy31415) [reply src/app/server/java/ChipFabricProvider-JNI.cpp:59]: Then add a comment for it: this does a 'GET' and then does not really use the result. This is non-obvious (and I still question if this is sane/valid ... why do we care at all?)

File: src/app/server/java/ChipFabricProvider-JNI.cpp (hunk 1 above)
Comment (bzbarsky-apple) [reply src/app/server/java/ChipFabricProvider-JNI.cpp]: Yes, that code does not need this check either.

Comment (bzbarsky-apple) [reply src/app/server/java/ChipFabricProvider-JNI.cpp]: I'll do a PR to fix the existing silliness.

Comment (bzbarsky-apple) [reply src/app/server/java/ChipFabricProvider-JNI.cpp]: https://github.com/project-chip/connectedhomeip/pull/18772

**Review by bzbarsky-apple — APPROVED** [review]

**Review by lazarkov — APPROVED** [review]

**Review by chrisdecenzo — APPROVED** [review]

**Pull request merged**